File: android/uikit/src/main/java/com/opensource/xt/uikit/XTUIViewController.kt
package com.opensource.xt.uikit

import android.app.Activity
import android.app.Fragment
import android.graphics.Color
import android.os.Handler
import android.view.View
import android.view.ViewGroup
import com.eclipsesource.v8.V8
import com.eclipsesource.v8.V8Array
import com.eclipsesource.v8.V8Object
import com.opensource.xt.core.*
import com.opensource.xt.uikit.libraries.keyboard.KeyboardHeightObserver
import com.opensource.xt.uikit.libraries.keyboard.KeyboardHeightProvider
import java.lang.ref.WeakReference

/**
 * Created by cuiminghui on 2017/9/5.
 */
open class XTUIViewController: XTUIFragment(), XTComponentInstance, KeyboardHeightObserver {

    override var objectUUID: String? = null
    lateinit var xtrContext: XTUIContext

    override var view: XTUIView? = null
        set(value) {
            if (field != null) { return }
            if (value == null) { return }
            field = value
            value.viewDelegate = WeakReference(this)
        }

    var parentViewController: WeakReference<XTUIViewController>? = null
        internal set

    var childViewControllers: List<XTUIViewController> = listOf()
        internal set

    open fun setContentView(activity: Activity) {
        XTUIScreen.resetScreenInfo(activity)
        this.requestFragment().let {
            val transaction = activity.fragmentManager.beginTransaction()
            transaction.replace(android.R.id.content, it)
            transaction.commit()
            it.setupKeyboardHeightProvider(activity)
        }
    }

    open fun attachFragment(activity: Activity, fragmentID: Int) {
        XTUIScreen.resetScreenInfo(activity)
        this.requestFragment().let {
            val transaction = activity.fragmentManager.beginTransaction()
            transaction.replace(fragmentID, it)
            transaction.commit()
            it.setupKeyboardHeightProvider(activity)
        }
    }

    fun scriptObject(): V8Object? {
        return xtrContext.evaluateScript("objectRefs['$objectUUID']") as? V8Object
    }

    open fun requestFragment(): XTUIViewController {
        return this
    }

    open fun viewDidLoad() {
        scriptObject()?.let {
            XTContext.invokeMethod(it, "viewDidLoad")
            XTContext.release(it)
        }
    }

    open fun viewWillAppear() {
        scriptObject()?.let {
            XTContext.invokeMethod(it, "viewWillAppear")
            XTContext.release(it)
        }
    }

    open fun viewDidAppear() {
        scriptObject()?.let {
            XTContext.invokeMethod(it, "viewDidAppear")
            XTContext.release(it)
        }
    }

    open fun viewWillDisappear() {
        scriptObject()?.let {
            XTContext.invokeMethod(it, "viewWillDisappear")
            XTContext.release(it)
        }
    }

    open fun viewDidDisappear() {
        scriptObject()?.let {
            XTContext.invokeMethod(it, "viewDidDisappear")
            XTContext.release(it)
        }
    }

    open fun viewWillLayoutSubviews() {
        scriptObject()?.let {
            XTContext.invokeMethod(it, "viewWillLayoutSubviews")
            XTContext.release(it)
        }
    }

    open fun viewDidLayoutSubviews() {
        scriptObject()?.let {
            XTContext.invokeMethod(it, "viewDidLayoutSubviews")
            XTContext.release(it)
        }
    }

    fun willMoveToParentViewController(parent: XTUIViewController?) {
        scriptObject()?.let {
            XTContext.invokeMethod(it, "_willMoveToParentViewController", listOf(
                    parent?.objectUUID ?: V8.getUndefined()
            ))
        }
    }

    fun didMoveToParentViewController(parent: XTUIViewController?) {
        scriptObject()?.let {
            XTContext.invokeMethod(it, "_didMoveToParentViewController", listOf(
                    parent?.objectUUID ?: V8.getUndefined()
            ))
        }
    }

    private var keyboardHeightProvider: KeyboardHeightProvider? = null

    private fun setupKeyboardHeightProvider(activity: Activity) {
        keyboardHeightProvider = KeyboardHeightProvider(activity)
        activity.findViewById(android.R.id.content).rootView?.post {
            keyboardHeightProvider?.start()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        keyboardHeightProvider?.close()
    }

    override fun onPause() {
        super.onPause()
        keyboardHeightProvider?.setKeyboardHeightObserver(null)
    }

    override fun onResume() {
        super.onResume()
        keyboardHeightProvider?.setKeyboardHeightObserver(this)
    }

    override fun onKeyboardHeightChanged(height: Int, orientation: Int) {
        this.lastKeyboardHeight = height
        handleKeyboardHeightChanged(height)
    }

    var lastKeyboardHeight = 0

    fun handleKeyboardHeightChanged(height: Int = this.lastKeyboardHeight) {
        scriptObject()?.let {
            if (height > 0) {
                XTContext.invokeMethod(it, "keyboardWillShow", listOf(
                        XTUIUtils.fromRect(XTUIRect(
                                0.0,
                                0.0,
                                this.view?.bounds?.width ?: 0.0,
                                height.toDouble() / resources.displayMetrics.density), xtrContext.runtime), 0.25))
            }
            else {
                XTContext.invokeMethod(it, "keyboardWillHide", listOf(0.25))
            }
            XTContext.release(it)
        }
    }

    class JSExports(val context: XTUIContext): XTComponentExport() {

        override val name: String = "_XTUIViewController"

        override fun exports(): V8Object {
            val exports = V8Object(context.runtime)
            exports.registerJavaMethod(this, "create", "create", arrayOf())
            exports.registerJavaMethod(this, "xtr_view", "xtr_view", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setView", "xtr_setView", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_parentViewController", "xtr_parentViewController", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_childViewControllers", "xtr_childViewControllers", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_addChildViewController", "xtr_addChildViewController", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_removeFromParentViewController", "xtr_removeFromParentViewController", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_navigationController", "xtr_navigationController", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_navigationBar", "xtr_navigationBar", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setNavigationBar", "xtr_setNavigationBar", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_showNavigationBar", "xtr_showNavigationBar", arrayOf(Boolean::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_hideNavigationBar", "xtr_hideNavigationBar", arrayOf(Boolean::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_showBackButton", "xtr_showBackButton", arrayOf(String::class.java))
            return exports
        }

        fun create(): String {
            val viewController = XTUIViewController()
            viewController.xtrContext = context
            val managedObject = XTManagedObject(viewController)
            viewController.objectUUID = managedObject.objectUUID
            XTMemoryManager.add(managedObject)
            return managedObject.objectUUID
        }

        fun xtr_view(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTUIViewController)?.view?.objectUUID
        }

        fun xtr_setView(viewRef: String, objectRef: String) {
            val view = XTMemoryManager.find(viewRef) as? XTUIView ?: return
            val viewController = XTMemoryManager.find(objectRef) as? XTUIViewController ?: return
            viewController.view = view
            view.post {
                viewController.viewDidLoad()
            }
        }

        fun xtr_parentViewController(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTUIViewController)?.parentViewController?.get()?.objectUUID
        }

        fun xtr_childViewControllers(objectRef: String): V8Array? {
            return (XTMemoryManager.find(objectRef) as? XTUIViewController)?.let {
                val v8Array = V8Array(context.runtime)
                it.childViewControllers.mapNotNull { it ->
                    return@mapNotNull it.objectUUID
                }.forEach { v8Array.push(it) }
                return@let v8Array
            }
        }

        fun xtr_addChildViewController(childControllerRef: String, objectRef: String) {
            val childController = XTMemoryManager.find(childControllerRef) as? XTUIViewController ?: return
            (XTMemoryManager.find(objectRef) as? XTUIViewController)?.let {
                if (childController.parentViewController == null) {
                    childController.willMoveToParentViewController(it)
                    it.childViewControllers.toMutableList()?.let { mutable ->
                        mutable.add(childController)
                        childController.parentViewController = WeakReference(it)
                        it.childViewControllers = mutable.toList()
                    }
                    childController.didMoveToParentViewController(it)
                }
            }
        }

        fun xtr_removeFromParentViewController(objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTUIViewController)?.let {
                it.parentViewController?.let { parentViewController ->
                    it.willMoveToParentViewController(null)
                    parentViewController.get()?.childViewControllers?.toMutableList()?.let { mutable ->
                        mutable.remove(it)
                        parentViewController.get()?.childViewControllers = mutable.toList()
                    }
                    it.didMoveToParentViewController(null)
                }
                it.parentViewController = null
            }
        }

        fun xtr_navigationController(objectRef: String): String? {
            (XTMemoryManager.find(objectRef) as? XTUIViewController)?.let {
                var currentParentViewController = it.parentViewController?.get()
                while (currentParentViewController != null) {
                    if (currentParentViewController is XTUINavigationController) {
                        return currentParentViewController.objectUUID
                    }
                    currentParentViewController = currentParentViewController.parentViewController?.get()
                }
            }
            return null
        }

        fun xtr_navigationBar(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTUIViewController)?.requestFragment()?.navigationBar?.objectUUID
        }

        fun xtr_setNavigationBar(navigationBarRef: String, objectRef: String) {
            val navigationBar = XTMemoryManager.find(navigationBarRef) as? XTUINavigationBar ?: return
            (XTMemoryManager.find(objectRef) as? XTUIViewController)?.requestFragment()?.navigationBar = navigationBar
        }

        fun xtr_showNavigationBar(value: Boolean, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTUIViewController)?.requestFragment()?.navigationBarHidden = false
        }

        fun xtr_hideNavigationBar(value: Boolean, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTUIViewController)?.requestFragment()?.navigationBarHidden = true
        }

        fun xtr_showBackButton(objectRef: String): Boolean {
            return (XTMemoryManager.find(objectRef) as? XTUIViewController)?.activity?.intent?.getBooleanExtra("XTUIShowBackButton", false) ?: false
        }

    }

}

File: ios/Sources/UIKit/XTUIViewController.m
//
//  XTUIViewController.m
//  XTSample
//
//  Created by 崔明辉 on 2017/8/25.
//  Copyright © 2017年 UED Center, YY Inc. All rights reserved.
//

#import "XTUIViewController.h"
#import "XTUIUtils.h"
#import "XTContext.h"
#import "XTUINavigationController.h"
#import "XTUINavigationBar.h"
#import "XTUIView.h"
#import <objc/runtime.h>
#import "XTMemoryManager.h"

@implementation UINavigationController(XTUIPatch)

- (BOOL)shouldAutorotate {
    return YES;
}

- (UIInterfaceOrientation)preferredInterfaceOrientationForPresentation {
    return self.childViewControllers.lastObject.preferredInterfaceOrientationForPresentation;
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
    return self.childViewControllers.lastObject.supportedInterfaceOrientations;
}

@end

@interface XTUIViewController ()

@property (nonatomic, assign) BOOL originalSettted;
@property (nonatomic, assign) UIStatusBarStyle originalStatusBarStyle;
@property (nonatomic, assign) BOOL originalNavigationBarHidden;
@property (nonatomic, strong) XTUINavigationBar *navigationBar;
@property (nonatomic, assign) BOOL navigationBarHidden;
@property (nonatomic, strong) UIView *innerView;
@property (nonatomic, weak) JSContext *context;
@property (nonatomic, strong) id keyboardWillShowObserver;
@property (nonatomic, strong) id keyboardWillHideObserver;

@end

@implementation XTUIViewController

+ (NSString *)name {
    return @"_XTUIViewController";
}

+ (NSString *)create {
    XTUIViewController *viewController = [XTUIViewController new];
    XTManagedObject *managedObject = [[XTManagedObject alloc] initWithObject:viewController];
    [XTMemoryManager add:managedObject];
    viewController.context = [JSContext currentContext];
    viewController.objectUUID = managedObject.objectUUID;
    return managedObject.objectUUID;
}


- (void)dealloc {
    JSValue *scriptObject = [self scriptObject];
    if (scriptObject != nil) {
        [scriptObject invokeMethod:@"dealloc" withArguments:nil];
    }
    [self unsetKeyboardNotifications];
#ifdef LOGDEALLOC
    NSLog(@"XTUIViewController dealloc.");
#endif
}


- (instancetype)init
{
    self = [super init];
    if (self) {
        _originalStatusBarStyle = [UIApplication sharedApplication].statusBarStyle;
        _navigationBarHidden = YES;
        self.automaticallyAdjustsScrollViewInsets = NO;
        [self setupKeyboardNotifications];
    }
    return self;
}

- (JSValue *)scriptObject {
    return [self.context evaluateScript:[NSString stringWithFormat:@"objectRefs['%@']", self.objectUUID]];
}

- (UIInterfaceOrientation)preferredInterfaceOrientationForPresentation {
    return UIInterfaceOrientationPortrait;
}

- (UIInterfaceOrientationMask)supportedInterfaceOrientations {
    if ([self scriptObject] != nil) {
        NSArray *supportOrientations = [self scriptObject][@"supportOrientations"].toArray;
        NSInteger value = 0;
        if ([supportOrientations containsObject:@(1)]) {
            value = value | UIInterfaceOrientationMaskPortrait;
        }
        if ([supportOrientations containsObject:@(2)]) {
            value = value | UIInterfaceOrientationMaskPortraitUpsideDown;
        }
        if ([supportOrientations containsObject:@(3)]) {
            value = value | UIInterfaceOrientationMaskLandscapeLeft;
        }
        if ([supportOrientations containsObject:@(4)]) {
            value = value | UIInterfaceOrientationMaskLandscapeRight;
        }
        return value;
    }
    return UIInterfaceOrientationMaskPortrait;
}

#pragma mark - Keyboard Notifications

- (void)setupKeyboardNotifications {
    static NSDictionary *lastUserInfo;
    __weak XTUIViewController *welf = self;
    self.keyboardWillShowObserver = [[NSNotificationCenter defaultCenter] addObserverForName:UIKeyboardWillShowNotification object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification * _Nonnull note) {
        __strong XTUIViewController *strongSelf = welf;
        JSValue *value = strongSelf.scriptObject;
        if (value) {
            NSDictionary *userInfo = note.userInfo ?: lastUserInfo;
            if (userInfo == nil) {
                return ;
            }
            lastUserInfo = userInfo;
            [value invokeMethod:@"keyboardWillShow" withArguments:@[
                                                                      [JSValue fromRect:[userInfo[UIKeyboardFrameEndUserInfoKey] CGRectValue]],
                                                                      userInfo[UIKeyboardAnimationDurationUserInfoKey] ?: @(0),
                                                                      ]];
        }
    }];
    self.keyboardWillHideObserver = [[NSNotificationCenter defaultCenter] addObserverForName:UIKeyboardWillHideNotification object:nil queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification * _Nonnull note) {
        __strong XTUIViewController *strongSelf = welf;
        JSValue *value = strongSelf.scriptObject;
        if (value) {
            [value invokeMethod:@"keyboardWillHide" withArguments:@[
                                                                      note.userInfo[UIKeyboardAnimationDurationUserInfoKey] ?: @(0),
                                                                      ]];
        }
    }];
}

- (void)unsetKeyboardNotifications {
    [[NSNotificationCenter defaultCenter] removeObserver:self.keyboardWillShowObserver];
    [[NSNotificationCenter defaultCenter] removeObserver:self.keyboardWillHideObserver];
}

+ (NSString *)xtr_view:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTUIViewController class]]) {
        XTUIView *view = (id)[(XTUIViewController *)obj innerView];
        if ([view isKindOfClass:[XTUIView class]]) {
            return view.objectUUID;
        }
    }
    else if ([obj isKindOfClass:[UIViewController class]]) {
        XTUIView *view = (id)obj.view;
        if ([view isKindOfClass:[XTUIView class]]) {
            return view.objectUUID;
        }
    }
    return nil;
}

+ (void)xtr_setView:(NSString *)viewRef objectRef:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    UIView *view = [XTMemoryManager find:viewRef];
    if ([obj isKindOfClass:[UIViewController class]] && [view isKindOfClass:[UIView class]]) {
        obj.view = [UIView new];
        obj.view.backgroundColor = view.backgroundColor ?: [UIColor whiteColor];
        [obj.view addSubview:view];
        if ([obj isKindOfClass:[XTUIViewController class]]) {
            [(XTUIViewController *)obj setInnerView:view];
        }
        [obj viewDidLoad];
    }
}

+ (NSDictionary *)xtr_safeAreaInsets:(NSString *)objectRef {
    XTUIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTUIViewController class]]) {
        CGFloat topLength = obj.topLayoutGuide.length;
        CGFloat bottomLength = obj.bottomLayoutGuide.length;
        if (@available(iOS 11.0, *)) {
            topLength = obj.view.safeAreaInsets.top;
            bottomLength = obj.view.safeAreaInsets.bottom;
        }
        if (obj.navigationBar != nil && !obj.navigationBarHidden) {
            topLength += 44.0;
        }
        if (!obj.navigationBar.translucent) {
            topLength = 0;
        }
        if (topLength == 0) {
            topLength = 20;
        }
        return [JSValue fromInsets:UIEdgeInsetsMake(topLength, 0, bottomLength, 0)];
    }
    return [JSValue fromInsets:UIEdgeInsetsZero];
}

+ (NSString *)xtr_parentViewController:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[UIViewController class]]) {
        id parentViewController = [obj parentViewController];
        if ([parentViewController conformsToProtocol:@protocol(XTComponent)]) {
            return [parentViewController objectUUID];
        }
    }
    return nil;
}

+ (NSArray<NSString *> *)xtr_childViewControllers:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    NSMutableArray *childViewControllers = [NSMutableArray array];
    if ([obj isKindOfClass:[UIViewController class]]) {
        for (id viewController in obj.childViewControllers) {
            if ([viewController conformsToProtocol:@protocol(XTComponent)]) {
                [childViewControllers addObject:[viewController objectUUID] ?: @""];
            }
        }
    }
    return [childViewControllers copy];
}

+ (void)xtr_addChildViewController:(NSString *)childControllerRef objectRef:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[UIViewController class]]) {
        UIViewController *childController = [XTMemoryManager find:childControllerRef];
        if ([childController isKindOfClass:[UIViewController class]]) {
            [obj addChildViewController:childController];
        }
    }
}

+ (void)xtr_removeFromParentViewController:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[UIViewController class]]) {
        [obj removeFromParentViewController];
    }
}

+ (NSString *)xtr_navigationController:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[UIViewController class]]) {
        UINavigationController *naviCtl = obj.navigationController;
        if ([naviCtl isKindOfClass:[XTUINavigationController class]]) {
            return [(XTUINavigationController *)naviCtl objectUUID];
        }
        else if ([naviCtl isKindOfClass:[UINavigationController class]]) {
            return [XTUINavigationController clone:naviCtl];
        }
    }
    return nil;
}

+ (void)xtr_presentViewController:(NSString *)viewControllerRef animated:(BOOL)animated objectRef:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    UIViewController *target = [XTMemoryManager find:viewControllerRef];
    if ([obj isKindOfClass:[UIViewController class]] && [target isKindOfClass:[UIViewController class]]) {
        [obj presentViewController:target animated:animated completion:nil];
    }
}

+ (void)xtr_dismissViewController:(BOOL)animated objectRef:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[UIViewController class]]) {
        [obj dismissViewControllerAnimated:animated completion:nil];
    }
}

#pragma mark - ViewController callbacks

- (void)loadView {
    self.view = [UIView new];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"loadView" withArguments:@[]];
        }
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.automaticallyAdjustsScrollViewInsets = NO;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewDidLoad" withArguments:@[]];
        }
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.navigationItem.hidesBackButton = YES;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewWillAppear" withArguments:@[]];
        }
    }
    if (self.navigationController != nil && !self.originalSettted) {
        self.originalSettted = YES;
        self.originalStatusBarStyle = [UIApplication sharedApplication].statusBarStyle;
        self.originalNavigationBarHidden = self.navigationController.navigationBar.hidden;
    }
    self.navigationController.navigationBar.hidden = YES;
    [[UIApplication sharedApplication] setStatusBarStyle:self.navigationBarLightContent ? UIStatusBarStyleLightContent : UIStatusBarStyleDefault];
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    self.navigationItem.hidesBackButton = NO;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewDidAppear" withArguments:@[]];
        }
    }
}

static UINavigationController *tmpNavigationController;
static BOOL onPanning;

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    self.navigationItem.hidesBackButton = YES;
    [UIApplication sharedApplication].statusBarStyle = self.originalStatusBarStyle;
    self.navigationController.navigationBar.hidden = self.originalNavigationBarHidden;
    tmpNavigationController = self.navigationController;
    if (!onPanning) {
        [self viewWillForceDisappear:animated];
    }
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewWillDisappear" withArguments:@[]];
        }
    }
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    self.navigationItem.hidesBackButton = NO;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewDidDisappear" withArguments:@[]];
        }
    }
    if (self.navigationController == nil && !self.originalNavigationBarHidden && tmpNavigationController != nil) {
        tmpNavigationController.navigationBar.alpha = 1.0;
    }
    tmpNavigationController = nil;
}

- (void)viewWillForceDisappear:(BOOL)animated {
    if (!tmpNavigationController.navigationBar.hidden && animated) {
        tmpNavigationController.navigationBar.alpha = 0.0;
        [UIView animateWithDuration:0.25 animations:^{
            tmpNavigationController.navigationBar.alpha = 1.0;
        }];
    }
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    if (self.navigationBar != nil && !self.navigationBarHidden) {
        CGFloat topLength = self.topLayoutGuide.length + 44.0;
        if (@available(iOS 11.0, *)) {
            topLength = self.view.safeAreaInsets.top + 44.0;
        }
        self.navigationBar.frame = CGRectMake(0, 0, self.view.bounds.size.width, topLength);
        if (self.navigationBar.translucent) {
            self.innerView.frame = self.view.bounds;
        }
        else {
            self.innerView.frame = CGRectMake(0,
                                              topLength,
                                              self.view.bounds.size.width,
                                              self.view.bounds.size.height - topLength);
        }
    }
    else {
        self.innerView.frame = self.view.bounds;
    }
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewWillLayoutSubviews" withArguments:@[]];
        }
    }
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewDidLayoutSubviews" withArguments:@[]];
        }
    }
}

- (void)willMoveToParentViewController:(id)parent {
    [super willMoveToParentViewController:parent];
    self.navigationBar.shouldShowBackBarButtonItem = [self.navigationController.childViewControllers indexOfObject:self] != 0;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"_willMoveToParentViewController"
                  withArguments:[parent conformsToProtocol:@protocol(XTComponent)] ? @[[parent objectUUID] ?: [NSNull null]] : @[]];
        }
    }
    if (parent == nil) {
        [self.navigationController.interactivePopGestureRecognizer
         addTarget:self
         action:@selector(onPopGesture:)];
    }
}

- (void)didMoveToParentViewController:(id)parent {
    [super didMoveToParentViewController:parent];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"_didMoveToParentViewController"
                  withArguments:[parent conformsToProtocol:@protocol(XTComponent)] ? @[[parent objectUUID] ?: [NSNull null]] : @[]];
        }
    }
    if (parent == nil) {
        [tmpNavigationController.interactivePopGestureRecognizer
         removeTarget:self
         action:@selector(onPopGesture:)];
    }
}

- (void)onPopGesture:(UIPanGestureRecognizer *)sender {
    if (sender.state == UIGestureRecognizerStateBegan) {
        onPanning = YES;
    }
    else if (sender.state == UIGestureRecognizerStateEnded) {
        onPanning = NO;
    }
    CGFloat progress = [sender locationInView:self.view.window].x / self.view.window.bounds.size.width;
    if (![tmpNavigationController.childViewControllers containsObject:self] && !self.originalNavigationBarHidden) {
        tmpNavigationController.navigationBar.alpha = progress;
    }
}

- (void)setNavigationBar:(XTUINavigationBar *)navigationBar {
    if (_navigationBar != nil) {
        [_navigationBar removeFromSuperview];
    }
    _navigationBar = navigationBar;
    if (_navigationBar != nil) {
        _navigationBar.viewController = self;
        [self.view addSubview:_navigationBar];
    }
    [self viewWillLayoutSubviews];
}

- (void)setTitle:(NSString *)title {
    [super setTitle:title];
    if (self.navigationBar) {
        self.navigationBar.innerItem.title = title;
    }
}

- (void)setNavigationBarLightContent:(BOOL)navigationBarLightContent {
    _navigationBarLightContent = navigationBarLightContent;
    [[UIApplication sharedApplication]
     setStatusBarStyle:_navigationBarLightContent ? UIStatusBarStyleLightContent : UIStatusBarStyleDefault];
    [self setNeedsStatusBarAppearanceUpdate];
}

- (UIStatusBarStyle)preferredStatusBarStyle {
    return self.navigationBarLightContent ? UIStatusBarStyleLightContent : UIStatusBarStyleDefault;
}

+ (NSString *)xtr_navigationBar:(NSString *)objectRef {
    XTUIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTUIViewController class]]) {
        return obj.navigationBar.objectUUID;
    }
    return nil;
}

+ (void)xtr_setNavigationBar:(NSString *)barRef objectRef:(NSString *)objectRef {
    XTUINavigationBar *bar = [XTMemoryManager find:barRef];
    XTUIViewController *obj = [XTMemoryManager find:objectRef];
    if ([bar isKindOfClass:[XTUINavigationBar class]] && [obj isKindOfClass:[XTUIViewController class]]) {
        obj.navigationBar = bar;
    }
}

+ (void)xtr_showNavigationBar:(BOOL)animated objectRef:(NSString *)objectRef {
    XTUIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTUIViewController class]]) {
        obj.navigationBarHidden = NO;
        obj.navigationBar.shouldShowBackBarButtonItem = [obj.navigationController.childViewControllers indexOfObject:obj] != 0;
        if (animated) {
            [UIView animateWithDuration:0.25 delay:0.0 usingSpringWithDamping:1.0 initialSpringVelocity:8.0 options:kNilOptions animations:^{
                obj.navigationBar.alpha = 1.0;
                [obj viewWillLayoutSubviews];
            } completion:nil];
        }
        else {
            obj.navigationBar.alpha = 1.0;
            [obj viewWillLayoutSubviews];
        }
    }
}

+ (void)xtr_hideNavigationBar:(BOOL)animated objectRef:(NSString *)objectRef {
    XTUIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTUIViewController class]]) {
        obj.navigationBarHidden = YES;
        if (animated) {
            [UIView animateWithDuration:0.25 delay:0.0 usingSpringWithDamping:1.0 initialSpringVelocity:8.0 options:kNilOptions animations:^{
                obj.navigationBar.alpha = 0.0;
                [obj viewWillLayoutSubviews];
            } completion:nil];
        }
        else {
            obj.navigationBar.alpha = 0.0;
            [obj viewWillLayoutSubviews];
        }
    }
}

@end
